# 1.온라인 선호지수 분석용 데이터 생성
## Search1과 Custom에 공통적으로 포함된 Client Id만으로 데이터프레임 생성

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pathlib import Path
import os

In [10]:
#데이터가져오기
Product = pd.read_csv("C:/Users/lsh93/Desktop/data/Product.csv")
Search1 = pd.read_csv("C:/Users/lsh93/Desktop/data/Search1.csv")
Search2 = pd.read_csv("C:/Users/lsh93/Desktop/data/Search2.csv")
Custom = pd.read_csv("C:/Users/lsh93/Desktop/data/Custom.csv")
Session = pd.read_csv("C:/Users/lsh93/Desktop/data/Session.csv")
Master = pd.read_csv("C:/Users/lsh93/Desktop/data/Master.csv")

C:\Users\lsh93\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3020: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [4]:
#구매데이터의 client 수 != 고객데이터의 client 수 
#구매데이터의 client 수 != 검색데이터의 client 수
print('Product shape: ',Product.shape)
print('Custom shape: ',Custom.shape)
print('Search1 shape: ',Search1.shape)
print('Session shape: ',Session.shape)
print()
print('구매에 참여한 client 수: ',len(Product['CLNT_ID'].value_counts()))
print('고객데이터에 있는 unique한 client 수: ',len(Custom['CLNT_ID'].value_counts()))
print('세션에 참여한 unique한 client 수: ',len(Session['CLNT_ID'].value_counts()))
print('검색에 참여한 client 수: ', len(Search1['CLNT_ID'].value_counts()))

Product shape:  (5024906, 8)
Custom shape:  (671679, 3)
Search1 shape:  (2884943, 4)
Session shape:  (2712907, 9)

구매에 참여한 client 수:  922737
고객데이터에 있는 unique한 client 수:  671679
세션에 참여한 unique한 client 수:  922737
검색에 참여한 client 수:  511477


- 세션에 참여한 client 수와 구매에 참여한 client 수는 동일하다 
- 상품을 구매했을지라도 검색을 하지 않은 client들이 있다 ) 922737 - 511477 만큼
- 고객데이터에 모든 client들의 정보가 있는 것은 아니다 

In [5]:
temp1 = pd.merge(Product,Session, how = 'inner' , on = 'CLNT_ID')

In [6]:
print(len(temp1['CLNT_ID'].value_counts()))

922737


In [6]:
temp2 = pd.merge(Custom, Product, how = 'right' , on = 'CLNT_ID')

In [7]:
print(len(temp1['CLNT_ID'].value_counts()))

922737


In [8]:
#null값 체크 --> 이건 unique한 건 아니지만 그래도 null값이 있다는 걸 의미함 
print(temp2['CLNT_AGE'].isnull().sum().sum())

1036319


# Product Outlier 제거
- 데이터 전처리 : outlier 제거
- HIT_SEQ가 1인 것 제거하기 

In [11]:
#HIT_SEQ가 1인 row 제거하기 (outlier로 간주하라고 하였음)
print('제거 전: ',len(Product['CLNT_ID'].value_counts()))  
idx = (Product['HITS_SEQ'] != 1) 
Product = Product[idx] 
print('제거 후: ',len(Product['CLNT_ID'].value_counts()))  

제거 전:  922737
제거 후:  922618


In [28]:
Product[Product['CLNT_ID'] == 4758889]  

CLNT_ID   SESS_ID  HITS_SEQ    PD_C         PD_ADD_NM PD_BRA_NM  \
533815   4758889  10607258        24  587541  색상:블랙|사이즈:F / 1개        비키   
597057   4758889  10607258       195  587541  색상:블랙|사이즈:F / 1개        비키   
655687   4758889  10607258       298  587541  색상:블랙|사이즈:F / 1개        비키   
661037   4758889  10730334        76  692624        색상:블랙 / 1개      LG전자   
4954520  4758889  10607258       298  776358       사이즈:28 / 1개     게스(진)   
5022807  4758889  10607258       195  776358       사이즈:28 / 1개     게스(진)   

        PD_BUY_AM PD_BUY_CT  
533815     45,000         1  
597057     45,000         1  
655687     45,000         1  
661037     45,340         1  
4954520    39,000         1  
5022807    39,000         1

## 1.1 상품 구매량 count를 위한 테이블 만들기

- 상품 구매 count를 하기 위해서, Product에서 HIT_SEQ = 1을 제거하고(앞에서 이미 제거됨), Search1과 custom에 둘 다 존재하는 Client 기준으로 정렬
- 한 client의 hit번호가 가장 큰 것에 대해서만 따로 빼서 매트릭스를 만들고, 그 후 상품분류에 대해 value_counts()해 주면 될듯

In [12]:
#Search1과 custom의 공통된 client id 뽑고, unique하게 만들어주기
Client = pd.merge(Search1[['CLNT_ID']], Custom[['CLNT_ID']])
Unique_Client = Client.drop_duplicates(subset=None, keep='last', inplace=False)

In [13]:
Unique_Client['CLNT_ID'].value_counts()

5461714    1
557637     1
5280327    1
5328413    1
2685514    1
6353483    1
4246092    1
4768333    1
55886      1
2150991    1
1583696    1
1581649    1
5767765    1
5773910    1
2365283    1
1079898    1
553563     1
5784157    1
6314590    1
6400608    1
2380969    1
2136646    1
3191363    1
4842089    1
2772219    1
4954662    1
2855463    1
2353704    1
1303081    1
5857480    1
          ..
4580630    1
2987288    1
2465049    1
4762753    1
6130971    1
1081954    1
4589347    1
5615932    1
4015362    1
4740254    1
1896362    1
1211621    1
6448358    1
1532795    1
5179014    1
3592504    1
4859118    1
3331312    1
5954801    1
6487285    1
5141990    1
4908278    1
6505990    1
2266360    1
5938425    1
6448371    1
4371708    1
3319038    1
6466815    1
6291456    1
Name: CLNT_ID, Length: 375864, dtype: int64

In [14]:
#Search1과 custom에 공통으로 있는 client id에 해당하는 구매내역만 dataframe으로 만들기 
Product_Client = pd.merge(Product, Unique_Client, how = 'inner', on = 'CLNT_ID')
print(len(Product_Client['CLNT_ID'].value_counts()))   

375864


In [115]:
#Product_Client = Product_Client.drop_duplicates(subset=None, keep='last', inplace=False)
#print(len(Product_Client['CLNT_ID'].value_counts()))   

375864


In [15]:
Product_Client[Product_Client['CLNT_ID']==4140076] 

CLNT_ID   SESS_ID  HITS_SEQ    PD_C                            PD_ADD_NM  \
0   4140076  10189797        13  570603     색상:BLK0_(BLK0)BLACK|사이즈:120 / 2개   
1   4140076  10342512        23  422358  선택:02.204143-02S|사이즈:M5W7(240) / 1개   
2   4140076   6526651        22  196272      색상:선택3)248(네이비)|사이즈:L(105) / 1개   
3   4140076   8326529       199  305965                     사이즈:115_3XL / 1개   
4   4140076   8891563        19  305965                     사이즈:115_3XL / 1개   
5   4140076   6526651       193  196272      색상:선택2)102(블랙)|사이즈:XL(110) / 1개   
6   4140076   6591360       149  196272      색상:선택2)102(블랙)|사이즈:XL(110) / 2개   
7   4140076   6526651        22  365025                   색상:WH|사이즈:165 / 1개   
8   4140076   4756252       147  794908                                   1개   
9   4140076  10342512        23  403243                          사이즈:M9 / 1개   
10  4140076   9147546       474  544957                                   1개   
11  4140076   6526651        22  364992                   색상:BG|사이즈:165 / 1개   
12  4140076   6591360       149  628713               사이즈:M7(260mm)_620 / 1개   
13  4140076  10574005       253  450131                  색상:BLK|사이즈:080 / 1개   
14  4140076   9147546       462  533783                     사이즈:XL(110) / 1개   
15  4140076   6526651        22  339321                   색상:BU|사이즈:165 / 1개   
16  4140076  10767546        19  600758                      사이즈:L(105) / 1개   
17  4140076  10767546        19  471477                         사이즈:165 / 1개   
18  4140076   7754761        71  315988     색상:DG(다크그레이)|사이즈:FREE(원사이즈) / 1개   
19  4140076   6517872       179  373461  품번:03.FITFLOP_J55_001_|사이즈:230 / 1개   
20  4140076   5338950        76  598881                                   1개   
21  4140076  10883079        29  689663                색상:그레이|직접입력1:240 / 1개   
22  4140076  10584304        50  689663                색상:그레이|직접입력1:235 / 1개   
23  4140076  10760242        26  689663                색상:그레이|직접입력1:240 / 1개   
24  4140076   6885554       134  288192                 색상:아이보리|사이즈:110 / 1개   
25  4140076   6885554       118  522552                색상:아이보리|Size:110 / 1개   
26  4140076   9992238       112  495110                         SIZE:XL / 1개   
27  4140076   8153697       123  555694             색상:화이트|사이즈:007(110) / 1개   
28  4140076   6591360       149  526369                         사이즈:XXL / 1개   
29  4140076  10342512        23  806382  색상:202356-159오이스터|사이즈:C06[155] / 1개   
30  4140076   4756252        76  570089                                   1개   
31  4140076   2051362        76  195687                      사이즈:L(105) / 1개   
32  4140076   6538217        26  739858                                   1개   
33  4140076   9899980        34  606095                           사이즈:3 / 1개   
34  4140076   7782189       125  606096                           사이즈:3 / 1개   
35  4140076  10574005       253  481952           구성:팬티3매|사이즈:80/11~12세 / 1개   
36  4140076   6526651       193  507637              색상:블랙(BLK)|사이즈:XXL / 1개   
37  4140076   6591360       149  507637              색상:블랙(BLK)|사이즈:XXL / 1개   
38  4140076   8608110       366  447887                     사이즈:120_4XL / 2개   
39  4140076   8235168       135  447887                     사이즈:120_4XL / 2개   
40  4140076   8891563        19  447887                     사이즈:120_4XL / 2개   
41  4140076   8326529       199  447887                     사이즈:120_4XL / 2개   
42  4140076   5384543        45  489476                                   1개   
43  4140076  10189797        13  411029   품번/사이즈:CY4574_회흰 / 120(XXXXL) / 1개   
44  4140076   6526651       193  262956                         사이즈:280 / 1개   
45  4140076   8608110       366  567645                     사이즈:115_3XL / 1개   
46  4140076   8235168       135  275704                        SIZE:3XL / 1개   
47  4140076   8235168       135  567645                     사이즈:115_3XL / 1개   
48  4140076   8326529       199  275704                        SIZE:3XL / 1개   
49  

## 1.2 Client Behavior 데이터 만들기

#### Client의 행동을 따라가보자
- Client가 어떻게 행동하는지 데이터 살펴보기
- Client Behavior 데이터 만들기 

- Client ID, 날짜(세션일자), 세션 ID, HIT번호, 검색키워드, 검색건수, 상품명, 구매건수 

- CLNT_ID(Product), SESS_DT(Session), SESS_ID(세션ID), HITS_SEQ(Product)
- KWD_NM(Search1), SEARCH_CNT(Search1), PD_NM(Master), PD_BUY_CT(Product)

#### Search1의 client < custom의 client < Product의 client = Session의 Client
#### Search1의 client 중에 custom 정보 없는 것 있다. 
#### 검색을 하고, custom 정보가 있는 client만을 기준으로 table 작성 (375864개의 client id)

In [17]:
Product_Client.head()

CLNT_ID   SESS_ID  HITS_SEQ    PD_C                            PD_ADD_NM  \
0  4140076  10189797        13  570603     색상:BLK0_(BLK0)BLACK|사이즈:120 / 2개   
1  4140076  10342512        23  422358  선택:02.204143-02S|사이즈:M5W7(240) / 1개   
2  4140076   6526651        22  196272      색상:선택3)248(네이비)|사이즈:L(105) / 1개   
3  4140076   8326529       199  305965                     사이즈:115_3XL / 1개   
4  4140076   8891563        19  305965                     사이즈:115_3XL / 1개   

  PD_BRA_NM PD_BUY_AM PD_BUY_CT  
0       데상트    39,000         2  
1       크록스    39,900         1  
2     프레드페리    47,600         1  
3  아디다스(의류)    46,700         1  
4  아디다스(의류)    46,700         1

In [18]:
data1 = pd.merge(Product_Client, Session)
print(len(data1['CLNT_ID'].value_counts()))

375864


In [23]:
#상품정보를 추가해준다
data2 = pd.merge(data1, Master, how = 'inner', on = 'PD_C')
print(len(data2['CLNT_ID'].value_counts()))

375864


In [43]:
#세션아이디도 동일한 것으로
#client -> 검색 -> 구매
Client_Behavior = pd.merge(data2, Search1, on=['CLNT_ID', 'SESS_ID'])
print(len(Client_Behavior['CLNT_ID'].value_counts()))
print('CLNT_ID랑 SESS_ID가 동일해야 하는 조건 때문에 client 수 약간 줄어들었음.')

375839
CLNT_ID랑 SESS_ID가 동일해야 하는 조건 때문에 client 수 약간 줄어들었음.


In [24]:
Client_Behavior

CLNT_ID   SESS_ID  HITS_SEQ    PD_C  \
0        4140076  10342512        23  422358   
1        4140076  10342512        23  403243   
2        4140076  10342512        23  806382   
3        4140076   6526651        22  196272   
4        4140076   6526651        22  196272   
5        4140076   6526651        22  196272   
6        4140076   6526651        22  196272   
7        4140076   6526651       193  196272   
8        4140076   6526651       193  196272   
9        4140076   6526651       193  196272   
10       4140076   6526651       193  196272   
11       4140076   6526651        22  365025   
12       4140076   6526651        22  365025   
13       4140076   6526651        22  365025   
14       4140076   6526651        22  365025   
15       4140076   6526651        22  364992   
16       4140076   6526651        22  364992   
17       4140076   6526651        22  364992   
18       4140076   6526651        22  364992   
19       4140076   6526651        22  339321   
20       4140076   6526651        22  339321   
21       4140076   6526651        22  339321   
22       4140076   6526651        22  339321   
23       4140076   6526651       193  507637   
24       4140076   6526651       193  507637   
25       4140076   6526651       193  507637   
26       4140076   6526651       193  507637   
27       4140076   6526651       193  262956   
28       4140076   6526651       193  262956   
29       4140076   6526651       193  262956   
...          ...       ...       ...     ...   
4883474  5010268   8741152        33  636556   
4883475  1082242   5661700        58  179401   
4883476  1082242   5661700        58  179401   
4883477   372204   5634118        48  403151   
4883478  3650646  10447349        33  563577   
4883479  5033434   4390462        75  507859   
4883480  3853640  10507265        54  687956   
4883481  3853640  10507265        54  687956   
4883482  4809147   7145157        63  224460   
4883483  4809147   7145157        63  224460   
4883484  4809147   7145157        63  224460   
4883485  4809147   7186972        47  224460   
4883486  5601778   8217108       130  423262   
4883487  5601778   8217108        91  423262   
4883488   901964      8750       285   15477   
4883489   901964      8750       285   15477   
4883490   901964      8750       285   15477   
4883491  1454857   9954443        26  816627   
4883492  1337534  10680863        53  419179   
4883493  4110447   5212278        90  636556   
4883494  4110447   6192376       114  391199   
4883495  4926545   9037714        54  455563   
4883496  5072758   8571308        27  819710   
4883497  2644758   6806773        21  388660   
4883498  5085969  10609581        48  590876   
4883499  5085969  10609581        48  590876   
4883500  5085969  10609581        48  590876   
4883501  3235159  10660304        94    3253   
4883502  4108900   2967180        27  642112   
4883503  4108900   2967180        56  642112   

                                   PD_ADD_NM     PD_BRA_NM PD_BUY_AM  \
0        선택:02.204143-02S|사이즈:M5W7(240) / 1개           크록스    39,900   
1                                사이즈:M9 / 1개           크록스    57,000   
2        색상:202356-159오이스터|사이즈:C06[155] / 1개           크록스    16,000   
3            색상:선택3)248(네이비)|사이즈:L(105) / 1개         프레드페리    47,600   
4            색상:선택3)248(네이비)|사이즈:L(105) / 1개         프레드페리    47,600   
5            색상:선택3)248(네이비)|사이즈:L(105) / 1개         프레드페리    47,600   
6            색상:선택3)248(네이비)|사이즈:L(105) / 1개         프레드페리    47,600   
7            색상:선택2)102(블랙)|사이즈:XL(110) / 1개         프레드페리    47,600   
8            색상:선택2)102(블랙)|사이즈:XL(110) / 1개         프레드페리    47,600   
9            색상:선택2)102(블랙)|사이즈:XL(110) / 1개         프레드페리    47,600   
10           색상:선택2)102(블랙)|사이즈:XL(110) / 1개         프레드페리    47,600   
11                        색상:WH|사이즈:165 / 1개         닥스 키즈    48,300   
12                        색상:WH|사이즈:165 / 1개         닥스 키즈    48,300   
13                      

- Hit번호만 다르고 나머지 element의 내용이 같은 row가 여러 개가 있는 경우(중복 행) -> Hit번호가 가장 큰 row만 남긴다 

# 2. 온라인 선호지수 개발

1) 구매선호지수(client i, 상품군 j)
- Product 데이터를 이용(실제 구매 데이터), 다른 item 대비 특정 상품군을 얼마나 구매하였는가  = 특정 상품군 구매 개수 / 각 client의 총 구매 개수
- client i = 1~375839 (Product, Search1, Custom모두에 정보가 있는 Client)

In [39]:
Client_Used = Client_Behavior[['CLNT_ID']]
Client_Used = Client_Used.drop_duplicates(subset=None, keep='last', inplace=False) 

In [40]:
print(len(Client_Used['CLNT_ID'].value_counts()))

375839


In [41]:
Product_Used = pd.merge(Product, Client_Used, how = 'inner', on = 'CLNT_ID')

In [42]:
print(len(Product_Used['CLNT_ID'].value_counts()))

375839
